### ISSUE TITLE
[cpp] [hl] Slow performance of trace command

### ISSUE THREAD

**Issue by sebthom** [HaxeFoundation/haxe#8830]
The following code runs significantly slower on HL and CPP compared to other targets:
```haxe
class TraceTest {

    public static function main() {
        var start = haxe.Timer.stamp();
        for (i in 1...2000)
            trace(i);
        var duration =  Math.round((haxe.Timer.stamp() - start)*100) / 100;
        trace('Duration: $duration seconds.');
    }
}
```

I tested Haxe 4RC5 on Windows 10, these are my results:

|Target|Duration|
|------|--------|
CPP| **18.73** secs
HL 1.10| **10.5** secs
PHP5| 0.88 secs
Java8| 0.85 secs
Node.js 10.16| 0.65 secs
C#| 0.51secs
PHP7| 0.51 secs
Neko2.2| 0.49 secs
Python3.7| 0.45 secs

**Comment by Aurel300** [comment]
Might have something to do with flushing stdout maybe?

**Comment by ncannasse** [comment]
Yes, HL will flush stdout on every print, maybe we could only do that when debugger is enabled

**Comment by sebthom** [comment]
> Yes, HL will flush stdout on every print, maybe we could only do that when debugger is enabled

@ncannasse The issue is still present in Haxe 4.2.5 and HL 1.1.2. Do you think you could implement your suggestion?

**Comment by yuxiaomao** [comment]
Maybe my computer is too fast for this, I'm getting `src/Main.hx:7: Duration: 0.01 seconds.` with latest HL (with or without debugger). However the last changes on `hl_sys_print` was from 5 years ago, I'm wondering if I'm missing something.
Could you please confirm that the issue is still present (or not)?

Edit: same result on try haxe. with haxe 4.3.6. so maybe it's fixed ?

**Comment by Simn** [comment]
We can't reproduce this with current Haxe + HL, and with the hxcpp problem being addressed by the linked PR this can be closed.

**Comment by lunacynne** [comment]
> Maybe my computer is too fast for this, I'm getting `src/Main.hx:7: Duration: 0.01 seconds.` with latest HL (with or without debugger). However the last changes on `hl_sys_print` was from 5 years ago, I'm wondering if I'm missing something.
> Could you please confirm that the issue is still present (or not)?
> 
> Edit: same result on try haxe. with haxe 4.3.6. so maybe it's fixed ?

weirdd
i tried this with hl today and got ~1.2s on average
latest haxe latest hl

ill investigate a bit later to see what the reason for the difference might be

**Comment by kLabz** [comment]
On what terminal were you running this?

**Comment by Simn** [comment]
> On what terminal were you running this?

Actually yeah, if I run this through cmd instead of vscode I get `Duration: 2.83 seconds.`. 

Edit: Though it should be noted that with hxcpp it goes from `Duration: 4.46 seconds.` before https://github.com/HaxeFoundation/hxcpp/pull/1199 to `Duration: 2.56 seconds.` after, so this may or may not be expected.

Edit 2: Then again, I get `source/Main.hx:7: Duration: 0.61 seconds.` with JVM, so I don't know.

**Comment by yuxiaomao** [comment]
I'm on Windows 11, and I have 0.01s in vscode, 0.01s in git bash, 0.38s in powershell, 0.47s in cmd. (HL/JIT)

**Comment by Simn** [comment]
> I'm on Windows 11, and I have 0.01s in vscode, 0.01s in git bash, 0.38s in powershell, 0.47s in cmd. (HL/JIT)

By itself that probably doesn't tell us much other than that Shiro has fast computers. Could you compare this to the hxcpp results?

**Comment by yuxiaomao** [comment]
cmd + hxcpp 4.3.2: 1.08s
cmd + hxcpp git: 0.50s
git bash + hxcpp 4.3.2: 0.02s
git bash + hxcpp git: 0.002s

I think the result is just similar between hl / hxcpp for me.

**Comment by sebthom** [comment]
I retested with Haxe 4.3.6 in a CMD window on Windows 10 22H2 (AMD Ryzen 7 PRO 7840U with 64GB RAM):
|Target|Duration|
|------|--------|
CPP (HXCPP 4.3.2 + MSVC 14.42)| **11.01** secs
CPP (HXCPP 4.3.74(git) + MSVC 14.42)| **7.4** secs
CPPIA (HXCPP 4.3.74(git) + MSVC 14.42)| **7.4** secs
HL 1.14| **8.03** secs
HL 1.15@82ea54c-win64 (nightly)| **8.54** secs
Java 21.0.2| 0.45 secs
C# (mono 6.12) | 0.53 secs
Lua 5.3.6 | 0.83 secs
Node.js 22.14.0| 0.89 secs
PHP 8.4.5| 0.7 secs
Neko 2.4| 0.39 secs
Python 3.13.2| 0.53 secs

So while there are some improvements since my tests six years ago there is still a considerable performance gap between CPP/HashLink and all other targets.
